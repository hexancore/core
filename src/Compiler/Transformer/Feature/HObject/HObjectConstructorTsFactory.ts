import ts from "typescript";
import type { HObjectPropertyTsMeta } from "./HObjectPropertyTsMeta";

export class HObjectToConstructorTsFactory {
  public create(properties: HObjectPropertyTsMeta[]): ts.ConstructorDeclaration {

    const callSuper = ts.factory.createExpressionStatement(ts.factory.createCallExpression(ts.factory.createSuper(), undefined, undefined));

    const propertiesSorted = [...properties.filter(p => !p.optional), ...properties.filter(p => p.optional)];
    const constructorParameters = propertiesSorted.map(p => this.createParameter(p));
    const constructorAssignments = propertiesSorted.map(p => this.createAssignment(p));

    return ts.factory.createConstructorDeclaration(
      [ts.factory.createModifier(ts.SyntaxKind.PublicKeyword)],
      constructorParameters,
      ts.factory.createBlock([callSuper, ...constructorAssignments], true)
    );
  }

  private createParameter(p: HObjectPropertyTsMeta): ts.ParameterDeclaration {
    return ts.factory.createParameterDeclaration(
      undefined,
      undefined,
      ts.factory.createIdentifier(p.name),
      p.optional ? ts.factory.createToken(ts.SyntaxKind.QuestionToken) : undefined,
      ts.factory.createKeywordTypeNode(ts.SyntaxKind.AnyKeyword),
      undefined
    );
  }

  private createAssignment(p: HObjectPropertyTsMeta): ts.Statement {
    return ts.factory.createExpressionStatement(
      ts.factory.createBinaryExpression(
        ts.factory.createPropertyAccessExpression(ts.factory.createThis(), p.name),
        ts.factory.createToken(ts.SyntaxKind.EqualsToken),
        ts.factory.createIdentifier(p.name)
      )
    );
  }
}