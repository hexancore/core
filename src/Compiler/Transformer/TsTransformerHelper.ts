import { Module } from "@nestjs/common";
import ts from "typescript";
import { TsImportHelper } from "./TsImportHelper";

export type ImportFromMapper = (importFrom: string) => string;

export type ImportedIdentifierMapper = (identifier: string) => ts.PropertyAccessChain | ts.Identifier;

export class TsTransfromerHelper {

  private static printer: ts.Printer;

  public static getLastImportIndex(source: ts.SourceFile): number {
    let lastIndex: number = 0;

    for (let i = 0; i < source.statements.length; ++i) {
      if (ts.isImportDeclaration(source.statements[i])) {
        lastIndex = i;
        continue;
      }

      if (ts.isDecorator(source.statements[i])) {
        break;
      }
    }

    return lastIndex;
  }

  public static transformDecoratorCallOfType(
    node: ts.Node,
    decoratorType: (...args) => any,
    updater: (decorator: ts.Decorator, decoratorExpression: ts.CallExpression) => ts.Node
  ): ts.Node {
    if (ts.isDecorator(node)) {
      const decoratorExpression = node.expression;
      if (ts.isCallExpression(decoratorExpression) && ts.isIdentifier(decoratorExpression.expression)) {
        if (decoratorExpression.expression.text === decoratorType.name) {
          return updater(node, decoratorExpression);
        }
      }
    }

    return node;
  }

  public static transformModuleDecoratorExpression(node: ts.Node, moduleMetaProvider: () => ({
    providers?: any[],
    imports?: ts.Expression[],
    exports?: any[],
  })): ts.Node {
    return TsTransfromerHelper.transformDecoratorCallOfType(node, Module, (decorator, decoratorExpression) => {
      const decoratorArguments = decoratorExpression.arguments;
      if (decoratorArguments.length > 0 && ts.isObjectLiteralExpression(decoratorArguments[0])) {
        const moduleMetaExpression = decoratorArguments[0];

        const assignedProps: string[] = [];
        const moduleMeta = moduleMetaProvider();
        const newProperties = moduleMetaExpression.properties.map(prop => {
          return TsTransfromerHelper.updatePropertyArrayAssigment(prop, (name: string) => {
            if (moduleMeta[name]) {
              assignedProps.push(name);
              return moduleMeta[name];
            }
            return null;
          });
        });

        for (const p in moduleMeta) {
          if (!assignedProps.includes(p)) {
            newProperties.push(ts.factory.createPropertyAssignment(p, ts.factory.createArrayLiteralExpression(moduleMeta[p])));
          }
        }
        const updatedModuleMetaExpression = ts.factory.updateObjectLiteralExpression(moduleMetaExpression, newProperties);
        const updatedDecoratorExpression = ts.factory.updateCallExpression(
          decoratorExpression,
          decoratorExpression.expression,
          decoratorExpression.typeArguments,
          [updatedModuleMetaExpression]
        );
        return ts.factory.updateDecorator(decorator, updatedDecoratorExpression);
      }

      return node;
    });
  }

  public static updatePropertyArrayAssigment(prop: ts.ObjectLiteralElementLike, newElementsProvider: (name: string) => ts.Expression[] | null): ts.ObjectLiteralElementLike {
    if (ts.isPropertyAssignment(prop) && ts.isIdentifier(prop.name)) {
      const newElements = newElementsProvider(prop.name.text);
      if (newElements) {
        if (ts.isArrayLiteralExpression(prop.initializer)) {
          const elements = [...prop.initializer.elements, ...newElements];
          return ts.factory.updatePropertyAssignment(prop, prop.name, ts.factory.updateArrayLiteralExpression(prop.initializer, elements));
        }
      }
    }

    return prop;
  }

  public static extractModuleSpecifierFromImportDeclaration(node: ts.ImportDeclaration): string {
    return (node.moduleSpecifier as ts.StringLiteral).text;
  }

  public static createNamedImportAccess(importDec: ts.ImportDeclaration, name: string, context: ts.TransformationContext): ts.PropertyAccessChain | ts.Identifier {
    const opts = context.getCompilerOptions();
    if (opts.module === ts.ModuleKind.Node16 || opts.module === ts.ModuleKind.CommonJS) {
      return ts.factory.createPropertyAccessChain(ts.factory.getGeneratedNameForNode(importDec), undefined, name);
    }

    return ts.factory.createIdentifier(name);

  }

  public static createImportDeclaration(importName: string | string[], importPath: string): ts.ImportDeclaration {
    importName = Array.isArray(importName) ? importName : [importName];

    const namedImports = ts.factory.createNamedImports(importName.map(n =>
      ts.factory.createImportSpecifier(false, undefined, ts.factory.createIdentifier(n)))
    );

    const importFrom = ts.factory.createStringLiteral(importPath);

    return ts.factory.createImportDeclaration(
      undefined,
      ts.factory.createImportClause(false, undefined, namedImports),
      importFrom
    );
  }

  /**
   * `let name = initializer`
   * `let name`
   * `let name: type`
   * `let name: type = initializer`
   */
  public static createLetStatement(name: string, type?: ts.TypeNode, initializer?: ts.Expression): ts.VariableStatement {
    return ts.factory.createVariableStatement(undefined,
      ts.factory.createVariableDeclarationList(
        [ts.factory.createVariableDeclaration(name, undefined, type, initializer)],
        ts.NodeFlags.Let,
      )
    );
  }

  public static createConstStatement(name: string, type?: ts.TypeNode, initializer?: ts.Expression): ts.VariableStatement {
    return ts.factory.createVariableStatement(undefined,
      ts.factory.createVariableDeclarationList(
        [ts.factory.createVariableDeclaration(name, undefined, type, initializer)],
        ts.NodeFlags.Const,
      )
    );
  }

  public static createImportFromMapper(sourceRoot: string): ImportFromMapper {
    const isInternalCoreTesting = sourceRoot.includes("/test/helper/libs");
    return isInternalCoreTesting ? (importFrom: string) => {
      return importFrom.replace('@hexancore/core', '@');
    } : (importFrom: string) => importFrom;
  }

  public static createImportHelper(sourceRoot: string, needFixImportAccess: boolean): TsImportHelper {
    const importFromMapper = this.createImportFromMapper(sourceRoot);
    return new TsImportHelper(importFromMapper, needFixImportAccess);
  }

  public static printFile(source: ts.SourceFile): string {
    if (!this.printer) {
      this.printer = ts.createPrinter({
        newLine: ts.NewLineKind.LineFeed,
        removeComments: false,
        omitTrailingSemicolon: true,
      });
    }

    return this.printer.printFile(source);
  }

  public static createGetEnumMemberFromUnknownValue(value: ts.Expression, enumType: ts.Expression | ts.EntityName): ts.ElementAccessExpression {
    // keyof typeof EnumType
    const keyofTypeofEnum = ts.factory.createTypeOperatorNode(ts.SyntaxKind.KeyOfKeyword, ts.factory.createTypeQueryNode(enumType as ts.EntityName));

    // v as keyof typeof EnumType
    const asExpression = ts.factory.createAsExpression(value, keyofTypeofEnum);

    // EnumType[v as keyof typeof EnumType]
    return ts.factory.createElementAccessExpression(
      enumType as ts.Expression,
      asExpression
    );
  }

  public static createDiagnostic(node: ts.Node, message: string): ts.Diagnostic {
    return {
      file: node.getSourceFile(),
      start: node.getStart(),
      length: node.getWidth(),
      messageText: message,
      category: ts.DiagnosticCategory.Error,
      code: 9001
    };
  }

  public static reportDiagnostics(diagnostics: ts.Diagnostic[], returnOutput = false): string {
    if (diagnostics.length === 0) {
      return '';
    }
    const formatHost: ts.FormatDiagnosticsHost = {
      getCanonicalFileName: fileName => fileName,
      getCurrentDirectory: ts.sys.getCurrentDirectory,
      getNewLine: () => ts.sys.newLine
    };

    const diagnosticMessages = ts.formatDiagnosticsWithColorAndContext(diagnostics, formatHost);
    if (returnOutput) {
      return diagnosticMessages;
    }

    console.error(diagnosticMessages);
    return diagnosticMessages;
  }
}