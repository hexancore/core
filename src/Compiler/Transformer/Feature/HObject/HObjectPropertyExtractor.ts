import ts from "typescript";
import { TsTransfromerHelper } from "../../TsTransformerHelper";
import { CollectionType, HObjectPropertyKind, HObjectPropertyTsMeta, HObjectPropertyPrimitiveType, ValueValidationRule } from "./HObjectPropertyTsMeta";

export class HObjectPropertyExtractor {
  public extract(node: ts.ClassDeclaration, sourceFile: ts.SourceFile, diagnostics: ts.Diagnostic[]): HObjectPropertyTsMeta[] {
    const properties: HObjectPropertyTsMeta[] = [];
    node.members.forEach(member => {
      if (ts.isPropertyDeclaration(member) && ts.isIdentifier(member.name)) {
        const meta = this.extractPropertyMeta(member, sourceFile, diagnostics);
        if (meta) {
          properties.push(meta);
        }
      }

      if (ts.isConstructorDeclaration(member)) {
        diagnostics.push(TsTransfromerHelper.createDiagnostic(node, "HObject user defined constructor is unsupported"));
        //this.collectFromConstructor(member, properties, sourceFile, diagnostics);
      }
    }, this);

    return properties;
  }

  private collectFromConstructor(node: ts.ConstructorDeclaration, properties: HObjectPropertyTsMeta[], sourceFile: ts.SourceFile, diagnostics: ts.Diagnostic[]): void {
    for (const param of node.parameters) {
      if (ts.isParameterPropertyDeclaration(param, node) && ts.isIdentifier(param.name)) {
        const meta = this.extractPropertyMeta(param, sourceFile, diagnostics);
        if (meta) {
          properties.push(meta);
        }
      }
    }
  }

  private extractPropertyMeta(member: ts.PropertyDeclaration | ts.ParameterPropertyDeclaration, sourceFile: ts.SourceFile, diagnostics: ts.Diagnostic[]) {
    if (member.type === undefined) {
      diagnostics.push(TsTransfromerHelper.createDiagnostic(member, 'Property has type "any", which is not allowed in HObject'));

      return undefined;
    }

    const typeMeta = this.extractPropertyType(member, member.type, sourceFile, diagnostics);
    if (!typeMeta) {
      return;
    }

    return new HObjectPropertyTsMeta(
      typeMeta.kind,
      (member.name as ts.Identifier).text,
      typeMeta.tsType,
      !!member.questionToken,
      typeMeta.validationRules,
      typeMeta.collectionType
    );
  }

  private extractPropertyType(
    parentNode: ts.Node,
    typeNode: ts.TypeNode,
    sourceFile: ts.SourceFile,
    diagnostics: ts.Diagnostic[]
  ): Pick<HObjectPropertyTsMeta, 'tsType' | 'kind' | 'validationRules' | 'collectionType'> | undefined {
    // v.int.between<10, 100>[] & v.items.exactly<10>
    // v.int[] & v.items.exactly<10>
    // string[] & v.items.exactly<10>
    if (ts.isIntersectionTypeNode(typeNode)) {
      return this.extractTypeFromIntesectionTypeNode(parentNode, typeNode, sourceFile, diagnostics);
    }

    // TODO: support type union

    // v.int.between<10, 100>[]
    // v.int[]
    // string[]
    if (ts.isArrayTypeNode(typeNode)) {
      return this.extractTypeFromArrayTypeNode(parentNode, typeNode, sourceFile, diagnostics);
    }

    // v.int.between<10, 100>
    // v.int
    // string
    return this.extractTypeFromNode(parentNode, typeNode, sourceFile, diagnostics);
  }

  private extractTypeFromIntesectionTypeNode(
    parentNode: ts.Node,
    typeNode: ts.IntersectionTypeNode,
    sourceFile: ts.SourceFile,
    diagnostics: ts.Diagnostic[]
  ): Pick<HObjectPropertyTsMeta, 'tsType' | 'kind' | 'validationRules' | 'collectionType'> | undefined {
    if (ts.isArrayTypeNode(typeNode.types[0])) {
      const itemType = this.extractTypeFromNode(parentNode, typeNode.types[0].elementType, sourceFile, diagnostics);
      if (!itemType) {
        return undefined;
      }

      const itemsRuleType = this.extractTypeFromNode(parentNode, typeNode.types[1], sourceFile, diagnostics);
      if (!itemsRuleType) {
        return itemType;
      }

      if (itemsRuleType.validationRules && itemsRuleType.validationRules[0].rule.startsWith('v.items')) {
        return {
          tsType: itemType.tsType,
          kind: itemType.kind,
          validationRules: [...itemType.validationRules!, itemsRuleType.validationRules[0]],
          collectionType: CollectionType.Array,
        };
      } else {
        diagnostics.push(TsTransfromerHelper.createDiagnostic(parentNode, 'Only v.items.* can be used for array HObject property'));
        return undefined;
      }
    }

    diagnostics.push(TsTransfromerHelper.createDiagnostic(parentNode, 'unsupported property type intersection definition'));
    return undefined;
  }

  private extractTypeFromArrayTypeNode(
    parentNode: ts.Node,
    typeNode: ts.ArrayTypeNode,
    sourceFile: ts.SourceFile,
    diagnostics: ts.Diagnostic[]
  ): Pick<HObjectPropertyTsMeta, 'tsType' | 'kind' | 'validationRules' | 'collectionType'> | undefined {
    const itemType = this.extractTypeFromNode(parentNode, typeNode.elementType, sourceFile, diagnostics);
    if (!itemType) {
      return undefined;
    }

    return {
      tsType: itemType.tsType,
      kind: itemType.kind,
      validationRules: itemType.validationRules,
      collectionType: CollectionType.Array,
    };
  }

  private extractTypeFromNode(
    parentNode: ts.Node,
    node: ts.TypeNode,
    sourceFile: ts.SourceFile,
    diagnostics: ts.Diagnostic[]
  ): Pick<HObjectPropertyTsMeta, 'tsType' | 'kind' | 'validationRules'> | undefined {
    const primitiveType = this.extractPrimitiveTypeFromNode(node);
    if (primitiveType) {
      return {
        kind: HObjectPropertyKind.PRIMITIVE,
        tsType: primitiveType,
      };
    }


    if (ts.isTypeReferenceNode(node)) {
      if (ts.isQualifiedName(node.typeName)) {
        const typeName = node.typeName.getText(sourceFile);
        if (typeName.startsWith('v.')) {
          const validationRule = this.extractValidationRuleFromNode(typeName, node, sourceFile);
          return {
            kind: HObjectPropertyKind.PRIMITIVE,
            tsType: validationRule.primitiveType,
            validationRules: [validationRule]
          };
        }
      } else {
        return {
          kind: HObjectPropertyKind.HOBJECT,
          tsType: node,
        };
      }
    }

    diagnostics.push(TsTransfromerHelper.createDiagnostic(parentNode, 'unsupported property type definition'));
    return undefined;
  }

  private extractPrimitiveTypeFromNode(node: ts.TypeNode): HObjectPropertyPrimitiveType | null {
    switch (node.kind) {
      case ts.SyntaxKind.StringKeyword:
        return HObjectPropertyPrimitiveType.string;
      case ts.SyntaxKind.BooleanKeyword:
        return HObjectPropertyPrimitiveType.boolean;
      case ts.SyntaxKind.NumberKeyword:
        return HObjectPropertyPrimitiveType.number;
      case ts.SyntaxKind.BigIntKeyword:
        return HObjectPropertyPrimitiveType.bigint;
    }
    return null;
  }

  private extractValidationRuleFromNode(typeName: string, node: ts.TypeReferenceNode, sourceFile: ts.SourceFile): ValueValidationRule {
    const args = node.typeArguments ? node.typeArguments.map(arg => arg.getText(sourceFile)) : ([] as string[]);
    return new ValueValidationRule(typeName, args);
  }
}