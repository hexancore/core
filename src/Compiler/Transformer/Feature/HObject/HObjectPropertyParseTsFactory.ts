import ts from "typescript";
import { CollectionType, HObjectPropertyPrimitiveType, HObjectPropertyTsMeta } from "./HObjectPropertyTsMeta";
import { TsTransfromerHelper } from "../../TsTransformerHelper";

export class HObjectPropertyParseTsFactory {
  public create(meta: HObjectPropertyTsMeta): ts.Statement[] {
    if (meta.isPrimitive()) {
      switch (meta.tsType) {
        case HObjectPropertyPrimitiveType.string: return this.createStringParse(meta);
      }
    } else {
      // TODO HObject parse create
      return [];
    }

    return [];
    //throw new LogicError('Unspported property meta in parse(), meta: ' + meta);

  }

  private createStringParse(meta: HObjectPropertyTsMeta): ts.Statement[] {
    /*
    let initializer;
    if (meta.validationRules) {
      if (meta.validationRules[0].isStringLengthRule()) {
        switch (meta.validationRules[0].extraRuleParts.join('.')) {
          case "length":
            initializer = this.createPrimitivePropertyParseCode(meta, "parseStringLength", {
              scalar: meta.validationRules[0].args[0],
            });
          break;
          case "length.min":
            initializer =
          break;
          case "length.max":
            initializer =
          break;
        }
      }
    } else {
      initializer = this.createPrimitivePropertyParseCode(meta, "parseString");
    }* */

    const initializer = this.createPrimitivePropertyParseCode(meta, "parseString");
    return meta.optional
      ? this.createOptionalPropertyParse(meta, initializer)
      : [TsTransfromerHelper.createConstStatement(meta.name, undefined, initializer)];
  }

  private createPrimitivePropertyParseCode(meta: HObjectPropertyTsMeta, helperMethod: string): ts.Expression {
    if (meta.collectionType === CollectionType.Array) {
      return this.createParsePrimitiveArray(meta, helperMethod);
    } else {
      return this.createPlainParseHelperCall(meta, helperMethod);
    }
  }

  private createOptionalPropertyParse(meta: HObjectPropertyTsMeta, initializer: ts.Expression): ts.Statement[] {
    const letPropertyVar = TsTransfromerHelper.createLetStatement(meta.name);
    // `p.propName === undefined`
    const ifCondition = ts.factory.createBinaryExpression(
      this.createPlainObjPropertyAccess(meta),
      ts.factory.createToken(ts.SyntaxKind.ExclamationEqualsEqualsToken),
      ts.factory.createIdentifier('undefined')
    );

    const optionalPropertyIf = ts.factory.createIfStatement(
      ifCondition,
      ts.factory.createBlock([this.createAssignInitializerToLetVar(meta, initializer)], true)
    );
    return [letPropertyVar, optionalPropertyIf];
  }

  private createAssignInitializerToLetVar(meta: HObjectPropertyTsMeta, initializer: ts.Expression): ts.Statement {
    return ts.factory.createExpressionStatement(
      ts.factory.createBinaryExpression(
        ts.factory.createIdentifier(meta.name),
        ts.factory.createToken(ts.SyntaxKind.EqualsToken),
        initializer,
      )
    );
  }

  private createParsePrimitiveArray(meta: HObjectPropertyTsMeta, parse: ts.Expression | string) {
    parse = typeof parse === 'string' ? this.createPlainParseHelperPropertyAccess(parse) : parse;
    return this.createPlainParseHelperCall(meta, "parsePrimitiveArray", [parse]);
  }

  private createPlainParseHelperCall(meta: HObjectPropertyTsMeta, method: string, args: ts.Expression[] = []) {
    return ts.factory.createCallExpression(
      this.createPlainParseHelperPropertyAccess(method),
      undefined,
      [
        this.createPlainObjPropertyAccess(meta),
        ...args,
        ts.factory.createStringLiteral(meta.name),
        ts.factory.createIdentifier('issues')
      ]
    );
  }

  private createPlainParseHelperPropertyAccess(name: string): ts.PropertyAccessExpression {
    return ts.factory.createPropertyAccessExpression(
      ts.factory.createIdentifier("PlainParseHelper"),
      ts.factory.createIdentifier(name)
    );
  }

  private createPlainObjPropertyAccess(meta: HObjectPropertyTsMeta): ts.PropertyAccessExpression {
    return ts.factory.createPropertyAccessExpression(
      this.createPlainObjIdentifier(),
      ts.factory.createIdentifier(meta.name)
    );
  }

  private createPlainObjIdentifier(): ts.Identifier {
    return ts.factory.createIdentifier("p");
  }



}