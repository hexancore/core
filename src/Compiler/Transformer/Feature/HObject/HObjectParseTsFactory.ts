
import ts from "typescript";
import type { HObjectPropertyTsMeta } from "./HObjectPropertyTsMeta";

import type { ImportDeclarationWrapper } from "../../Helper/ImportDeclarationWrapper";
import { TsTransfromerHelper } from "../../TsTransformerHelper";
import { HObjectPropertyParseTsFactory } from "./HObjectPropertyParseTsFactory";
import { HObjectKind, type FeatureHObjectMeta } from "@/Util/Feature/Meta";

const HOBJECT_ANY_TYPE_MAP = {
  [HObjectKind.Command]: {
    anyTypeName: "AnyHCommand",
    typeName: "HCommandType"
  },
  [HObjectKind.Query]: {
    anyTypeName: "AnyHQuery",
    typeName: "HQueryType"
  },
  [HObjectKind.Event]: {
    anyTypeName: "AnyHEvent",
    typeName: "HEventType"
  },
  [HObjectKind.Dto]: {
    anyTypeName: "AnyDto",
    typeName: "HDtoType"
  },
  [HObjectKind.ValueObject]: {
    anyTypeName: "AnyValueObject",
    typeName: "HValueObjectType"
  },
  [HObjectKind.Entity]: {
    anyTypeName: "AnyEntity",
    typeName: "HEntityType"
  },
  [HObjectKind.AggregateRoot]: {
    anyTypeName: "AnyAggregateRoot",
    typeName: "HAggregateRootType"
  }
};

export class HObjectParseTsFactory {
  private propertyTsFactory: HObjectPropertyParseTsFactory;

  public constructor() {
    this.propertyTsFactory = new HObjectPropertyParseTsFactory();
  }

  //`public static parse<T extends AnyDto>(this: DtoType<T>, plain: unknown): R<T, PlainParseError>`
  public create(meta: FeatureHObjectMeta, hObjectClassName: string, properties: HObjectPropertyTsMeta[], hCommonImportDecl: ImportDeclarationWrapper): ts.MethodDeclaration {
    const returnType = ts.factory.createTypeReferenceNode(
      hCommonImportDecl.getEntityName('R'), [ts.factory.createTypeReferenceNode(hObjectClassName, [])]
    );

    const typeInfo = HOBJECT_ANY_TYPE_MAP[meta.kind];

    const parameters = [
      ts.factory.createParameterDeclaration(
        undefined,
        undefined,
        ts.factory.createIdentifier('this'),
        undefined,
        ts.factory.createTypeReferenceNode(typeInfo.typeName, [ts.factory.createTypeReferenceNode("T")])
      ),
      ts.factory.createParameterDeclaration(undefined, undefined, 'plain', undefined, ts.factory.createKeywordTypeNode(ts.SyntaxKind.UnknownKeyword))
    ];

    const genericT = ts.factory.createTypeParameterDeclaration(
      undefined,
      ts.factory.createIdentifier("T"),
      ts.factory.createTypeReferenceNode(ts.factory.createIdentifier(typeInfo.anyTypeName), undefined)
    );

    return ts.factory.createMethodDeclaration(
      [ts.factory.createModifier(ts.SyntaxKind.PublicKeyword), ts.factory.createModifier(ts.SyntaxKind.StaticKeyword)],
      undefined, // *
      'parse',
      undefined, // '?'
      [genericT],
      parameters,
      returnType,
      ts.factory.createBlock(this.createBody(hObjectClassName, properties, hCommonImportDecl), true)
    );
  }

  private createBody(hObjectClassName: string, properties: HObjectPropertyTsMeta[], hCommonImportDecl: ImportDeclarationWrapper): ts.Statement[] {
    properties = [...properties.filter(p => !p.optional), ...properties.filter(p => p.optional)];

    const statements: ts.Statement[] = [
      this.createCheckIsObject(hObjectClassName, hCommonImportDecl),
      this.createPlainObjVarDeclaration(hObjectClassName),
      this.createIssuesVarDeclaration(),
    ];

    const propNames: string[] = [];
    for (const p of properties) {
      statements.push(...this.propertyTsFactory.create(p, hCommonImportDecl));
      propNames.push(p.name);
    }

    statements.push(this.createCheckAnyIssues(hObjectClassName, hCommonImportDecl));
    statements.push(this.createReturnNew(hObjectClassName, hCommonImportDecl, propNames));

    return statements;
  }

  /**
    `
      if (typeof plain !== 'object') {
      return PlainParseHelper.HObjectIsNotObjectParseErr(hObjectClassName as any, plain]);
      }
    `
  */
  private createCheckIsObject(hObjectClassName: string, hCommonImportDecl: ImportDeclarationWrapper): ts.Statement {
    // `typeof plain !== 'object'`
    const condtionExpression = ts.factory.createBinaryExpression(
      ts.factory.createTypeOfExpression(ts.factory.createIdentifier('plain')),
      ts.factory.createToken(ts.SyntaxKind.ExclamationEqualsEqualsToken),
      ts.factory.createStringLiteral('object')
    );

    const block = ts.factory.createBlock([
      ts.factory.createReturnStatement(
        ts.factory.createCallExpression(
          ts.factory.createPropertyAccessExpression(
            hCommonImportDecl.get('PlainParseHelper'),
            'HObjectIsNotObjectParseErr'
          ),
          undefined,
          [
            this.createAsAny(hObjectClassName),
            ts.factory.createIdentifier('plain')
          ]
        )
      )
    ], true);

    return ts.factory.createIfStatement(condtionExpression, block);
  }

  // `const p = plain as Record<keyof TestDto, unknown>;`
  private createPlainObjVarDeclaration(hObjectClassName: string): ts.Statement {
    const initializer = ts.factory.createAsExpression(
      ts.factory.createIdentifier('plain'),
      ts.factory.createTypeReferenceNode(
        ts.factory.createIdentifier('Record'),
        [
          ts.factory.createTypeOperatorNode(ts.SyntaxKind.KeyOfKeyword, ts.factory.createTypeReferenceNode(hObjectClassName, undefined)),
          ts.factory.createKeywordTypeNode(ts.SyntaxKind.UnknownKeyword),
        ]
      )
    );
    return TsTransfromerHelper.createConstStatement("p", undefined, initializer);
  }

  // `const issues: PlainParseIssue[] = [];`
  private createIssuesVarDeclaration() {
    const type = ts.factory.createArrayTypeNode(ts.factory.createTypeReferenceNode('PlainParseIssue', undefined));
    const initializer = ts.factory.createArrayLiteralExpression([], false);
    return TsTransfromerHelper.createConstStatement("issues", type, initializer);

  }

  private createCheckAnyIssues(hObjectClassName: string, hCommonImportDecl: ImportDeclarationWrapper): ts.Statement {
    return ts.factory.createIfStatement(
      ts.factory.createBinaryExpression(
        ts.factory.createPropertyAccessExpression(
          this.issuesVarIdentifier(),
          ts.factory.createIdentifier('length')
        ),
        ts.factory.createToken(ts.SyntaxKind.GreaterThanToken),
        ts.factory.createNumericLiteral('0')
      ),
      ts.factory.createBlock(
        [
          ts.factory.createReturnStatement(
            ts.factory.createCallExpression(
              ts.factory.createPropertyAccessExpression(
                hCommonImportDecl.get('PlainParseHelper'),
                ts.factory.createIdentifier('HObjectParseErr')
              ),
              undefined,
              [
                this.createAsAny(hObjectClassName),
                this.issuesVarIdentifier(),
              ]
            )
          )
        ],
        true
      ),
      undefined
    );
  }

  private createReturnNew(hObjectClassName: string, hCommonImportDecl: ImportDeclarationWrapper, propNames: string[]) {
    return ts.factory.createReturnStatement(
      ts.factory.createAsExpression(
        ts.factory.createCallExpression(
          hCommonImportDecl.get("OK"),
          undefined,
          [
            ts.factory.createNewExpression(
              hCommonImportDecl.get(hObjectClassName),
              undefined,
              propNames.map((name) => this.createAsAny(name))
            )
          ]
        ),
        ts.factory.createKeywordTypeNode(ts.SyntaxKind.AnyKeyword)
      )
    );
  }

  private issuesVarIdentifier(): ts.Identifier {
    return ts.factory.createIdentifier("issues");
  }

  private createAsAny(name: string): ts.Expression {
    return ts.factory.createAsExpression(
      ts.factory.createIdentifier(name),
      ts.factory.createKeywordTypeNode(ts.SyntaxKind.AnyKeyword)
    );
  }

}