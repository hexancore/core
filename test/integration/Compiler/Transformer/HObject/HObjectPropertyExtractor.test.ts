/**
 * @group integration
 */

import { TsTransformerTestHelper } from "@/Compiler/Test/TsTransformerTestHelper";
import ts from "typescript";
import { TsTransfromerHelper } from "@/Compiler/Transformer/TsTransformerHelper";
import { LIBS_DIRNAME } from "@test/libs";
import type { HObjectPropertyTsMeta } from "@/Compiler/Transformer/Feature/HObject/HObjectPropertyTsMeta";
import { HObjectPropertyExtractor } from "@/Compiler/Transformer/Feature/HObject/HObjectPropertyExtractor";

describe(HObjectPropertyExtractor.constructor.name, () => {
  const helper = TsTransformerTestHelper.createFromTsConfig(process.cwd() + "/tsconfig.json");
  let extractor: HObjectPropertyExtractor;

  beforeAll(() => {
    extractor = new HObjectPropertyExtractor();
  });

  test("extract", () => {
    const sourceFilePath = LIBS_DIRNAME + "/test-lib/src/Book/Application/Book/Dto/TestTransformDto.ts";
    const diagnostics: ts.Diagnostic[] = [];
    let properties!: HObjectPropertyTsMeta[];
    const transformer = (source) => {
      ts.visitNodes(source.statements, (node) => {
        if (ts.isClassDeclaration(node)) {
          properties = extractor.extract(node,source, diagnostics);
        }
        return node;
      });

      return source;
    };

    helper.transpileModule(transformer, sourceFilePath);

    expect(properties.map(p => p.toString())).toMatchSnapshot();
    expect(TsTransfromerHelper.reportDiagnostics(diagnostics, true)).toEqual('');
  });

});


