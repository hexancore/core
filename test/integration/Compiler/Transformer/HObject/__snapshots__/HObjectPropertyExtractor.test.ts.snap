// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Function extract 1`] = `
[
  "optionalField?: string",
  "numberField!: number",
  "stringField!: string",
  "booleanField!: boolean",
  "bigintField!: bigint",
  "primitiveArrayField!: string[]",
  "uintField!: v.uint",
  "ruleWithArgsField!: v.int.between<-10,100>",
  "ruleArrayField!: v.int.between<-10,100>[]",
  "ruleArrayWithItemsField!: v.int.between<-10,100>[] & v.items.between<2,5>",
  "hObjField!: RefId",
  "hObjArrayField!: RefId[]",
]
`;
